src/plugins/dataflow-tool/components/dataflow-program.tsx

```diff
@@ -603,24 +603,26 @@ export class DataflowProgram extends BaseComponent<IProps, IState> {
     }
 
     const rect = this.getBoundingRectOfNodes();
+
     if (rect?.isValid) {
-      // Handle program too large for client
-      const newZoom = Math.min(k * clientWidth / (rect.right + margin),
-                               k * clientHeight / (rect.bottom + margin));
-      if (newZoom < k && rect.right > 0 && newZoom > 0) {
+      const widthQ = k * clientWidth / (rect.right + margin);
+      const heightQ = k * clientHeight / (rect.bottom + margin);
+      const newZoom = Math.min(widthQ, heightQ);
+
+      const tooSmall = rect.width < (clientWidth * .25) || rect.height < (clientHeight * .25);
+      const tooBig = newZoom < k && rect.right > 0 && newZoom > 0;
+
+      if (tooBig) {
         this.programEditor.view.area.transform = {k: newZoom, x: transform.x, y: transform.y};
         this.programEditor.view.area.update();
         return;
       }
 
-      // Handle program too small for client
-      const targetPercentage = .9;
-      if (rect.width < clientWidth * targetPercentage
-        || rect.height < clientHeight * targetPercentage) {
-          const newerZoom = Math.min(k * clientWidth * targetPercentage / rect.width,
-            k * clientHeight * targetPercentage / rect.height);
-          this.programEditor.view.area.transform = {k: newerZoom, x: transform.x, y: transform.y};
-          this.programEditor.view.area.update();
+      // If the program is way too small for the current work area, reset zoom to 1
+      // checking both to avoid flicker-inducing race condition
+      if ( tooSmall && !tooBig ) {
+        this.programEditor.view.area.transform = {k: 1, x: transform.x, y: transform.y};
```

Comment:
Would it work to make `k` slightly smaller than 1, maybe .95 or something? That could help with the flicker as well, I don't know.

Reply:
Seems to work...(in next commit)

Reply:
If I constrained in both directions I could still get a flicker, but it's very hard to arrive at that case now with below changes.